src/index.ml

```diff
@@ -296,6 +298,83 @@ struct
           Int64.of_float rounded
       end)
 
+  let try_load_log t path =
+    Log.debug (fun l ->
+        l "[%s] checking on-disk %s file" (Filename.basename t.root)
+          (Filename.basename path));
+    if Sys.file_exists path then (
+      let io =
+        IO.v ~fresh:false ~readonly:true ~generation:0L ~fan_size:0L path
+      in
+      let mem = Tbl.create 0 in
+      iter_io (fun e -> Tbl.replace mem e.key e.value) io;
+      Some { io; mem } )
+    else None
+
+  let sync_log t =
+    Log.debug (fun l ->
+        l "[%s] checking for changes on disk" (Filename.basename t.root));
+    let no_changes () =
+      Log.debug (fun l ->
+          l "[%s] no changes detected" (Filename.basename t.root))
+    in
+    let add_log_entry log e = Tbl.replace log.mem e.key e.value in
```

Comment:
Same question with `replace` vs. `add`

Reply:
I changed `replace` with `add`, but there is still one place where we still want to use `replace` (https://github.com/mirage/index/pull/175/files#diff-732d381615307f94d510e2dbdee74f4aR371). When we add new values, we want to replace old binding with new ones.

Reply:
What's the semantics that we want? yes, `Tbl.add` will keep multiple entries for the keys, but `Tbl.find` will always return the most recent one, which is exactly what we want I think. We just need to be careful with duplicated entries if we do a `Tbl.iter` later on. I don't know if it's a good idea or not, but `replace` can be very costly when the hashtbl buckets starts to be full. (probably something to benchmarks to see if it's worth it)

Reply:
Note: I'm fine to do these changes in a separate PR, as it has nothing todo with the explicit `ro_sync`

Reply:
I think `iter` is the only place (in the RO instance) where we care about duplicate entries, but we say that we do not guarantee this https://github.com/mirage/index/blob/master/src/index.mli#L176

I think we should have this changes in a different PR as the RW instance also uses `Tbl.replace` in a few places. I'll revert these changes in this PR.

Reply:
Please bear in mind that whenever a `merge` happens, this table folded into an array, sorted, and used to fill the new index file. At that point if you have multiple bindings for a key they will all be folded and written to disk during the merge, and you can retrieve them by reading the index, which is not what we want.

Reply:
> I'll revert these changes in this PR.

Done. (I'll open a replace/add PR).